bundles/org.openhab.binding.pentair/src/main/java/org/openhab/binding/pentair/internal/handler/PentairControllerHandler.java

```diff
@@ -0,0 +1,1134 @@
+    public static @Nullable PentairControllerHandler onlineController;
+    public boolean serviceMode = false;
+    public boolean uom = false;
+
+    public boolean syncTime = true;
+
+    private final Logger logger = LoggerFactory.getLogger(PentairControllerHandler.class);
+
+    protected @Nullable ScheduledFuture<?> syncTimeJob;
+
+    private int preambleByte = -1; // Byte to use after 0xA5 in communicating to controller. Not sure why this changes,
+                                   // but it requires to be in sync and up-to-date
+    private long lastScheduleTypeWrite;
+
+    private static final int CACHE_EXPIRY = (int) TimeUnit.SECONDS.toMillis(10);
+    private static final int CACHE_EXPIRY_LONG = (int) TimeUnit.MINUTES.toMillis(10);
+
+    protected final ExpiringCache<PentairControllerStatus> controllerStatusCache = new ExpiringCache<>(CACHE_EXPIRY);
+    protected final ExpiringCache<PentairHeatStatus> heatStatusCache = new ExpiringCache<>(CACHE_EXPIRY);
+
+    private int majorrev, minorrev;
+
+    protected final ExpiringCache<PentairControllerCircuit>[] circuitsCache = new ExpiringCache[NUM_CIRCUITS];
+    protected final ExpiringCache<PentairControllerSchedule>[] schedulesCache = new ExpiringCache[NUM_SCHEDULES];
+
+    protected @Nullable LightMode lightMode;
+
+    public PentairControllerHandler(Thing thing) {
+        super(thing);
+
+        for (int i = 0; i < NUM_SCHEDULES; i++) {
+            schedulesCache[i] = new ExpiringCache<>(CACHE_EXPIRY_LONG);
+        }
+
+        for (int i = 0; i < NUM_CIRCUITS; i++) {
+            circuitsCache[i] = new ExpiringCache<>(CACHE_EXPIRY_LONG);
+        }
+    }
+
+    @Override
+    public void readConfiguration() {
+        PentairControllerHandlerConfig config = getConfigAs(PentairControllerHandlerConfig.class);
+
+        this.id = config.id;
+        this.syncTime = config.synctime;
+    }
+
+    @Override
+    public void goOnline() {
+        // Only a single controller is supported on the Pentair bus so prevent multiple controller
+        // things being created.
+        if (onlineController != null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Another controller is already configured.");
+        } else {
+            super.goOnline();
+        }
+    }
+
+    @Override
+    public void finishOnline() {
+        super.finishOnline();
+
+        // setup syncTimeJob to run once a day, initial time to sync is 3 minutes after controller goes online. This is
+        // to prevent collision with main thread queries on initial startup
+        syncTimeJob = scheduler.scheduleWithFixedDelay(() -> syncTime(), 3, 24 * 60 * 60, TimeUnit.MINUTES);
+
+        scheduler.execute(() -> readControllerSettings());
+    }
+
+    public void syncTime() {
+        boolean synctime = ((boolean) getConfig().get("synctime"));
```

Comment:
Configuration parameters should be static final constants. Please make "synctime" a field in one of your constants classes.

Reply:
since the user can set this value via UI, can this really be "final"?

Reply:
I meant turn it into a constant field.

So in your PentairBindingConstants class (or whatever constants class you choose to put it in).
```
public static final String CONFIG_SYNCTIME = "synctime";
```

```suggestion
        boolean synctime = ((boolean) getConfig().get(CONFIG_SYNCTIME));
```